Step2: 문자열 계산기

mannue opened the pull request:
### class 역할
- Separator
  - static class 이며 differentiate 함수를 통하여 입력 값을 구분자로 구별하여 String 배열 값을 반환한다.
  - 입력값이 null 이거나 size 가 0 이면 RuntimeException 예외를 발생시킨다.

- StringAddCalculator
    - static class 이며 splitAndSum 함수를 가지고 있으며 String 값을 인수로 받아서 int 값을 반환한다.

- WholeNumber
  - VO 이며 , 정수 값만 가질수 있다.
  - 정수 가 아닌 다른 값이 입력으로 들어오면 RuntimeException 예외를 throw 한다.
  - add 메소드를 통해서 WholeNumber 끼리 더하기를 할수 있다.
  - of 메소드를 통해서 int 값으로 변환할수 있다.
 
- WholeNumbers 
  - sum 메소드를 통해서 WholeNumber 들을 계산하여 int 값을 반환한다.


바쁘신 시간에 리뷰 해 주셔서 감사합니다.
오늘 하루도 좋은 하루 되세요 👍

mannue committed:
docs(StringAddCalculator): 요구 사항 정리 와 class 역할 정리

mannue committed:
feat(WholeNumber): WholeNumber class 생성

WholeNumber 는 VO 객체이며 정수 값을 입력 받는다 .

mannue committed:
test(WholeNumber): WholeNumber 의 입력 값에  String 숫자 이외에 값 인 경우 에러를 발생한다.

mannue committed:
test(WholeNumber): WholeNumber 의 입력 값은 음수가 될수 없다.

mannue committed:
refactor(WholeNumber): 불필요한 생성자 제거 와 테스트 입력값 변경

mannue committed:
docs(StringAddCalculator): WholeNumber 역할 추가 정리

mannue committed:
feat(WholeNumber): add 메소드를 이용하여 WholeNumber 끼리 더하기

mannue committed:
refactor(WholeNumber): 음수일때 Exception class 변경

mannue committed:
feat(WholeNumber): of 메소드를 통해서 현재 상태  값을 반환한다.

mannue committed:
feat(WholeNumbers): WholeNumbers class 생성

mannue committed:
feat(WholeNumbers): sum 메서드를 이용하여 WholeNumber 들의 합 값을 가져온다.

mannue committed:
refactor(WholeNumbers): sum Test 를 위한 케이스 추가 와 생성 테스트 케이스 삭제

mannue committed:
feat(Separator): Separator , SeparatorTest class 생성 과 심플 성공 테스트 추가

mannue committed:
test(Separator): SeparatorTest 에 커스텀 구분자를 제외한 기본 구분자로 분리하기

mannue committed:
docs(Separator): 입력값이 잘못된 경우 문서 추가

mannue committed:
feat(Separator): custom separator 를 이용하여 분리

mannue committed:
feat(Separator): 입력값이 잚못된 경우 RuntimeException 이 발생한다.

mannue committed:
refactor(WholeNumbers): WholeNumber 의 add 를 사용하도록 수정

mannue committed:
feat(WholeNumbers): String[] 를 위한 생성자 추가

mannue committed:
feat(StringAddCalculator): StringAddCalculator 와 StringAddCalculatorTest 추가

mannue committed:
docs(docs): WholeNumberGenerator 삭제

seok-2-o commented on src/main/java/camp/nextstep/edu/step2/Separator.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.regex.Pattern;
+
+public class Separator {
+    private static final String regex = "//(.)\n(.*)";
```
상수는 대문자로 작성하는 것이 관례입니다.

seok-2-o replied:
사용하지 않는 상수를 선언하신 이유가 궁금하네요.

mannue replied:
상수는 대문자로 작성하는 것이 관례입니다. 
-> 네 대문자가 맞는것 같습니다. 세세한 코드 리뷰 감사합니다 👍 

사용하지 않는 상수를 선언하신 이유가 궁금하네요.
-> 향후 변경 여부를 위해서 static final 로 추출했는데 코드상에서 변경 하지 않았네요 감사합니다 . 
    정규식이 변경 가능성이 있어서 따로 추출하였습니다.

seok-2-o commented on src/main/java/camp/nextstep/edu/step2/Separator.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    private static Optional<String[]> splitByCustomDelimiter(String input) {
+        Matcher m = Pattern.compile("//(.)\n(.*)").matcher(input);
```
[Effective Java 3/E] 아이템6. 불필요한 객체 생성을 피하라

패턴은 생성에 드는 리소스가 많은 객체에 속하는데 이를 만들어놓고 사용한다면 어떤 장점이 있을까요?

mannue replied:
네 Separator 멤버 변수로 Pattern.comiile()  이부분만 추출 하고 메소드 안에서는 matcher 를 사용하는데 
객체를 계속 사용할수 있을것 같습니다. 
조언 감사합니다 👍

seok-2-o commented on src/main/java/camp/nextstep/edu/step2/WholeNumber.java:
```diff
@@ -0,0 +1,44 @@
+        this.number = wholeIntegerBy(input);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        WholeNumber that = (WholeNumber) o;
+        return number == that.number;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(number);
+    }
```
equals & hashCode 코드는 코드의 맨 하단에 위치하는 것ㅇ ㅣ좋습니다.

mannue replied:
private 보다 위에  위치했는데 equals & hashCode 보다 다른 부분이 더 중요할수 있겠네요 
코드 작성 시 하단으로 위치 하도록 노력하겠습니다 👍 
가르쳐 주셔서 감사합니다 ~

seok-2-o commented on src/main/java/camp/nextstep/edu/step2/WholeNumber.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    private WholeNumber add(int number) {
+        return new WholeNumber(String.valueOf(this.number + number));
```
```suggestion
        if(nuber == 0) {
            return this;
        }
        return new WholeNumber(String.valueOf(this.number + number));
```

seok-2-o replied:
소소하지만 자주 사용하는 패턴이니 알아두시면 좋을 거 ㅅ같아요.

mannue replied:
앗 감사합니다. 0 이면 굳이 연산할 필요가 없겠네요 
좋은 가르침 감사합니다 👍

seok-2-o commented on src/main/java/camp/nextstep/edu/step2/WholeNumber.java:
```diff
@@ -0,0 +1,44 @@
+        return wholeNumber.add(this.number);
+    }
+
+    private WholeNumber add(int number) {
```
33라인에서 수행해도 될 것 같은데 구지 나눠주신 이유가 궁금하네요.

mannue replied:
33라인에서 바로 접근시 입력 객체에서 다이렉트로 number 로 접근하기 때문에 위 코드처럼 나눴습니다.
가능한 직접 접근은 자제 하려고 합니다 ^^;; (자바지기님께서 직접 접근은 자제하라고 동영상에서 본적 있어요...)

seok-2-o replied:
다른 객체에 직접 접근하는것을 자제하는 것이 좋다는말이긴하지만 말씀하신 부분에도 일리가 있네요 👍

mannue replied:
의견 들어주셔서 감사합니다 👍

seok-2-o commented on src/main/java/camp/nextstep/edu/step2/WholeNumber.java:
```diff
@@ -0,0 +1,44 @@
+        return new WholeNumber(String.valueOf(this.number + number));
+    }
+
+    int of() {
```
의미를 명확하게 표현할 수 있는 메소드 명을 사용해보는 것은 어떤가요?

mannue replied:
해당 부분 추가 할때 getter 같은 느낌이라서 만들고 싶지 않았는데 제 생각으로는 좋은 방법이 없었네요 ㅠㅠ
이름은 WholeNumber.value() 밖에 생각이 안나네요 ㅠㅠ 
해당 이름으로 수정 후 아이디어나 조언 , 방법 등등 알려주시면 감사히 배우도록 하겠습니다 ㅠㅠ 
부족 해서 죄송합니다.

seok-2-o replied:
일반적으으로는 변경하신 value 를 많이 사용하는 것 같아요
getter를 사용하지 않는 것이 좋으나 일반적으로 모든 개발자들이 getter는 외부의 값을 내부로 꺼낸다라는
의미를 알기때문에 필요하다면 과감하게 사용해주세요. 

남이 읽기 좋은 코드를 고민하는 것이 클린코드의 시작점이라는 생각이 드네요 :)

mannue replied:
맞습니다. 남들이 읽기 좋은 코드를 만들어야 된다고 저도 생각합니다 :)
늦은 시간인데 감사합니다 ^^

seok-2-o commented on src/main/java/camp/nextstep/edu/step2/Separator.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    private static void validation(final String input) {
+        if (Objects.isNull(input) || input.isEmpty()) {
```
StringAddCalculator 7번 라인에서 수행되고 있는데 중복해서 수행되어야 할까요?

mannue replied:
SringAddCalculator 가 아닌 다른곳에서 Separator 를 사용했을때 잘못된 인자를 넣을수 있다고 생각해서 추가하였는데요
저도 이부분을 항상 고민하는데요 상위에서 체크한 부분을 클래스 안에서 다시 한번 체크 해야되는가 ? 라는 고민을 항상 하고 있는데요
@seok-2-o  이부분 에 대해서 조언 해주시면 감사하겠습니다 ^^

seok-2-o replied:
매번 체크하면 당연하게도 코드의 안정성은 올라가지만 타협해야하는 부분이 있다고 생각해요.
지금은 무조건 Separator를 거치게 코드가 작성되어있읜 한곳에서만 유효성 검증을 진행해도 괜찮지 않을까요?

mannue replied:
향후 별도로 사용될때 구현해도 되겠네요 
때로는 한곳에서 밖에 사용안할수도 있겠네요 
미리 확장 하지 않도록 해봐야겠네요 감사합니다 👍

seok-2-o reviewed (changes requested):
첫번째 미션 잘 진행해주셨네요 👍 
세세한 객체분리가 인상적이였습니다.

몇가지 코멘트 남겨두었으니 한번 고민해보면 좋을 것 같아요.

mannue committed:
refactor(Separator): REGEX 이름 을 대문자로 변경 과 Pattern 을 한번만 생성하도록 수정

mannue committed:
refactor(WholeNumber): 코드 리뷰 comment 수정

- eqauls 와 hashCode 는 중요도가 낮기때문에 하단 위치 하는게 좋다
- of 이름 보다는 명확하게 value 로
수정
- add 시 입력 값이 0 인 경우는 자신을 return

mannue commented:
바쁘신 가운데 리뷰 감사합니다. 👍 
조언 주신 부분을 수정 하였으며 주신 리뷰에 comment 추가 하였습니다.
가르침과 조언에 감사드립니다 😄

seok-2-o reviewed (approved):
피드백 빠르게 잘 반영해주셨네요 👍 
인사드릴때 말씀드린것처럼 제가 드린 의견이 정답은 아니니
꼭 모든 내용을 수정할 필요는 없다는 부분 다시 한번 말씀드려요.

본격적으로 로또 미션 진행해보아요 🔥

mannue reviewed (commented):
늦은 시간까지 꼼꼼하게 리뷰 봐주시고 조언 해주셔서 감사합니다 😄 
건강도 잘 챙기세요 ^^ 감사합니다 👍
